### ISSUE TITLE
[Performance] Store State always returns a new reference

### ISSUE THREAD

**Issue by justrhysism** [wbkd/react-flow#681]
The `useStoreState(state => state.nodes)` returns a new reference on every render. I am trying to implement grouping behaviour where a node needs to access "sibling" nodes.

 _"That's fine, I can just do a `nodes.filter()` and memoize it..."_

Except because the reference is new on every render, the memoization runs on every render :(.

I understand that this may be due to the underlying state management, but for now I'm having to use crude workarounds to not have it update on every render.

---

I might have to open a separate issue for this, but it would be even _better_ if the Elements were stored in a `Map` with index by the provided `id`. This would when fetching specific nodes we don't need to loop through the entire list each and every time _a la:`nodes.find(n => n.id === id)`_.

If there's anything I can do to assist with this, please let me know. Willing to assist and contribute under guidance.

**Comment by moklick** [comment]
Hey @justrhysism 

thanks for the input! We are also not super happy with the element state handling right now. 
We will look into other possibilities and if a Map works for us.

**Comment by justrhysism** [comment]
@moklick I've had a couple of iterations with grouping. Originally I was manually shifting each node with `updateNodePosDiff`. See: https://codesandbox.io/s/magical-heisenberg-nty4i 

However, that broke when upgrading to v7, so I switched to a (much simpler) method of adding all the nodes of the group to the selection via `addSelectedElements` when the group is "selected".

With respect to _how_ I was grouping, I just pass (via `data`) a `parentId` for the "parent" node, and the "parent" node has an array of ids for it's children.

---

The other feature I've had to disable for performance reasons is the resizing of the group node (see above codesandbox, and drag around the nodes within a group). As the tree grew, the number of loops, and the _length_ of the loops got longer resulting in some sort of Big O notated issue I can't be bothered to figure out—but it slowed down a lot.

Obviously I'm pushing the boundaries outside of the original design intentions, so for now I've just disabled some of those features.

**Comment by moklick** [comment]
Hey! 

This looks pretty impressive! We will think about how we can make it easier to implement things like custom grouping.
Do you have an idea what you need in order to simplify the grouping?

We changed the `updateNodePosDiff` so that all selected elements get updated (we need that for the multi selection).

**Comment by justrhysism** [comment]
@moklick to be honest, I don't really have an opinion on an "easy" grouping method. I am quite happy to handle it myself, but yeah as stated on the outset, the issue I have is with the new array reference on every render. I thought it might be just because it's computed, but unfortunately, when I switched to using the `elements` directly I had the same issue.

---

> We changed the updateNodePosDiff so that all selected elements get updated (we need that for the multi selection).

Yeah that makes sense. As I mentioned, I found a workaround by adding all the "siblings" to the selected elements so the group is moved "natively" with respect to the library.

It was a little annoying, but it _was_ a breaking change haha, so all good 😄
